model_signing/serialization/itemized.py

```diff
@@ -0,0 +1,228 @@
+        """
+        self._hasher_factory = hasher_factory
+        self._max_workers = max_workers
+
+    @override
+    def serialize(self, model_path: pathlib.Path) -> manifest.FileLevelManifest:
+        # TODO: github.com/sigstore/model-transparency/issues/196 - Add checks
+        # to exclude symlinks if desired.
+        dfs.check_file_or_directory(model_path)
+
+        paths = []
+        if model_path.is_file():
+            paths.append(model_path)
+        else:
+            # TODO: github.com/sigstore/model-transparency/issues/200 - When
+            # Python3.12 is the minimum supported version, this can be replaced
+            # with `pathlib.Path.walk` for a clearer interface, and some speed
+            # improvement.
+            for path in model_path.glob("**/*"):
+                dfs.check_file_or_directory(path)
+                if path.is_file():
+                    paths.append(path)
+
+        manifest_items = []
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self._max_workers
+        ) as tpe:
+            futures = [
+                tpe.submit(self._compute_hash, model_path, path)
+                for path in paths
+            ]
+            for future in concurrent.futures.as_completed(futures):
+                manifest_items.append(future.result())
+
+        return self._build_manifest(manifest_items)
+
+    def _compute_hash(
+        self, model_path: pathlib.Path, path: pathlib.Path
+    ) -> manifest.FileManifestItem:
+        """Produces the manifest item of the file given by `path`.
+
+        Args:
+            model_path: The path to the model.
+            path: Path to the file in the model, that is currently transformed
+              to a manifest item.
+
+        Returns:
+            The itemized manifest.
+        """
+        # Note: we need to force a PosixPath to canonicalize the manifest
+        relative_path = pathlib.PurePosixPath(path.relative_to(model_path))
+        digest = self._hasher_factory(path).compute()
+        return manifest.FileManifestItem(path=relative_path, digest=digest)
+
+    def _build_manifest(
+        self, items: Iterable[manifest.FileManifestItem]
+    ) -> manifest.FileLevelManifest:
+        """Builds an itemized manifest from a given list of items.
+
+        Every subclass needs to implement this method to determine the format of
+        the manifest.
+        """
+        return manifest.FileLevelManifest(items)
+
+
+class ShardedFilesSerializer(serialization.Serializer):
+    """Model serializers that produces an itemized manifest, at shard level.
+
+    Traverses the model directory and creates digests for every file found,
+    sharding the file in equal shards and computing the digests in parallel.
+
+    Since the manifest lists each item individually, this will also enable
+    support for incremental updates (to be added later).
+    """
+
+    def __init__(
+        self,
+        hasher_factory: Callable[
```

Comment:
nit: Rename this to `sharded_hasher_factory` and the above one to `file_hasher_factory` to reduce confusion.